### PR TITLE
Fixed exception when user tries to log in

### LINKED ISSUE #35952: Code: 1001. DB::Exception: std::out_of_range: unordered_map::at: key not found
Hi
When I used clickhouse-client to connect to the database, the database reported the following error!
<Error> TCPHandler: Code: 1001. DB::Exception: std::out_of_range: unordered_map::at: key not found. (STD_EXCEPTION), Stack trace (when copying this message, always i
nclude the lines below):

0. std::logic_error::logic_error(char const*) @ 0x18dd2a6d in ?
1. ? @ 0xa50afc9 in /usr/bin/clickhouse
2. ? @ 0xa50af80 in /usr/bin/clickhouse
3. DB::SettingsProfilesInfo::getProfileNames() const @ 0x1473a1b1 in /usr/bin/clickhouse
4. DB::SessionLog::addLoginSuccess(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<c
har> > >, DB::Context const&) @ 0x14739ab6 in /usr/bin/clickhouse
5. DB::Session::makeQueryContextImpl(DB::ClientInfo const*, DB::ClientInfo*) const @ 0x1472f947 in /usr/bin/clickhouse
6. DB::TCPHandler::receiveQuery() @ 0x1544e3ba in /usr/bin/clickhouse
7. DB::TCPHandler::receivePacket() @ 0x154477d6 in /usr/bin/clickhouse
8. DB::TCPHandler::runImpl() @ 0x15440dfe in /usr/bin/clickhouse
9. DB::TCPHandler::run() @ 0x15450e79 in /usr/bin/clickhouse
10. Poco::Net::TCPServerConnection::start() @ 0x164b264f in /usr/bin/clickhouse
11. Poco::Net::TCPServerDispatcher::run() @ 0x164b4aa1 in /usr/bin/clickhouse
12. Poco::PooledThread::run() @ 0x16671e49 in /usr/bin/clickhouse
13. Poco::ThreadImpl::runnableEntry(void*) @ 0x1666f1a0 in /usr/bin/clickhouse
14. ? @ 0x7fb387437609 in ?
15. __clone @ 0x7fb38735c163 in ?

Installed latest version clickhouse-server and clickhouse-client.
There was no such error in the version 20.9.2.20

### PR DESCRIPTION
**Pull request by Enmk**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed unable to log in (because of failure to create session_log entry) in rare case of messed up setting profiles.
...

Closes: #35952

Fixed `SettingsProfilesInfo::getProfileNames()` throwing `std::exception`
- Using profiles_with_implicit for profile IDs, since it seems to be better synchronized with names
- Trying harder: attempting to get name from `AccessControl` if it is missing from local cached names
- Throwing Exception if everything else fails

Fix is based on a stack trace of #35952 and some notes in bug description.
There are no tests added since I wasn't able to reproduce bug on my side, however same code path must be checked by existing tests.

### Bug analysis
- While populating `SettingsProfilesInfo::profiles_with_implicit` and `SettingsProfilesInfo::names_of_profiles`, `SettingsProfilesCache::substituteProfiles` can't find a profile by UUID in `SettingsProfilesCache::all_profiles`.
- So `SettingsProfilesInfo::profiles_with_implicit` and `SettingsProfilesInfo::names_of_profiles` do not have an entry for a profile from `SettingsProfilesInfo::profiles`.
- This could happen only if `SettingsProfilesCache::all_profiles` updated asynchronously, but it is protected by mutex.

So just to be safe `SettingsProfilesInfo::getProfileNames` uses profile `UID`s from `SettingsProfilesInfo::profiles_with_implicit` instead of `SettingsProfilesInfo::profiles`, since `SettingsProfilesInfo::profiles_with_implicit` and `SettingsProfilesInfo::names_of_profiles` are explicitly synchronized by `SettingsProfilesCache::substituteProfiles`.

### CODE REVIEW

**Comment by rschu1ze** [comment]
@Enmk I am not really familiar with the code and its constraints and I will certainly not oppose your fix. But, to be honest, it feels like a bandaid for an issue that should never have happened in the first place. I agree with your section "bug analysis", in particular that the member variables of `SettingsProfileInfo` are out-of-sync and that they are set (at least as far as I see) by `SettingsProfilesCache::substituteProfiles()`. So the question is what could have caused this problem. Thoughts:
- We could use Clang TSA to safeguard the locking in SettingsProfileCache, see #38068
- Is `substituteProfiles` really exception-safe? E.g. it does lots of `.insert()`, `.push_back()` and `.emplace()` where a bad_alloc exception could mess up the state of `SettingsProfileInfo`. This could be fixed by first reserving sufficient space in all members and then populating them.
- Maybe there are other places in this function which can throw.

Even if there is no repro (which is okay, perhaps the issue occurs really super sporadically), my preference would be to make the settings profile creating code more robust.

**Comment by Enmk** [comment]
> * Is `substituteProfiles` really exception-safe? E.g. it does lots of `.insert()`, `.push_back()` and `.emplace()` where a bad_alloc exception could mess up the state of `SettingsProfileInfo`. This could be fixed by first reserving sufficient space in all members and then populating them.

`substituteProfiles` is not exception safe, but there should be no way to modify inconsistently `SettingsProfilesInfo` in `SettingsProfilesInfo` due to exception. It only modifies the state of the stack-based `SettingsProfilesInfo` instance, so if an exception is thrown, said instance is not added to the `SettingsProfilesCache::profile_infos_cache` and lost.

From what I see, `substituteProfiles` is always called with a locked `SettingsProfilesCache::mutex` (directly or indirectly):
```mermaid
graph TD
    mergeSettingsAndConstraints --> mergeSettingsAndConstraintsFor --> substituteProfiles
    getSettingsProfileInfo --> |locks mutex | substituteProfiles
    getEnabledSettings --> |locks mutex | mergeSettingsAndConstraintsFor
    profileRemoved --> |locks mutex | mergeSettingsAndConstraints
    profileAddedOrChanged --> |locks mutex | mergeSettingsAndConstraints
```

Took another look, and probably fixed the most probable cause is `mergeSettingsAndConstraintsFor`, see 
https://github.com/ClickHouse/ClickHouse/pull/42641/commits/aa78f1c8421aa78b7a7c6de53d4ea8d235bd5263

@rschu1ze ^^

File: src/Access/SettingsProfilesInfo.cpp [hunk 1]
```diff
@@ -55,4 +62,26 @@ SettingsProfilesInfo::getConstraintsAndProfileIDs(const std::shared_ptr<const Se
...
+        if (p == names_of_profiles.end())
+        {
+            if (const auto name = access_control.tryReadName(profile_id))
+                // We could've updated cache here, but it is a very rare case, so don't bother.
+                result.push_back(*name);
+            else
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Unable to get profile name for {}", toString(profile_id));
+        }
+
+        result.push_back(p->second);
```
Comment (rschu1ze) [inline src/Access/SettingsProfilesInfo.cpp:81]: If `profile-id` is not found in `names_of_profiles` but in `access_control` (l. 76), then we fallthrough to l. 81 where the beyond-the-end iterator is dereferenced. The code should wrap l. 81 in an `else`. Also suggest to handle the common case first, i.e. check for `p != names_of_profiles.end()` in l. 72.

**Comment by rschu1ze** [comment]
Thanks for the analysis, I agree with it, and sorry for the delay (just got back from vacation)

Please find one smaller comment attached. Functional test `test_function_current_profiles` looks fishy, like to check?

**Comment by rschu1ze** [comment]
test_function_current_profiles unfortunately still fails.
```
[gw0] [100%] FAILED test_settings_profile/test.py::test_function_current_profiles 

=================================== FAILURES ===================================
________________________ test_function_current_profiles ________________________
[gw0] linux -- Python 3.8.10 /usr/bin/python3

    def test_function_current_profiles():
        instance.query("CREATE SETTINGS PROFILE P1, P2")
        instance.query("ALTER USER robin SETTINGS PROFILE P1, P2")
        instance.query("CREATE SETTINGS PROFILE P3 TO robin")
        instance.query("CREATE SETTINGS PROFILE P4")
        instance.query("CREATE SETTINGS PROFILE P5 SETTINGS INHERIT P4")
        instance.query("CREATE ROLE worker SETTINGS PROFILE P5")
        instance.query("GRANT worker TO robin")
        instance.query("CREATE SETTINGS PROFILE P6")
    
        session_id = new_session_id()
>       assert (
            instance.http_query(
                "SELECT defaultProfiles(), currentProfiles(), enabledProfiles()",
                user="robin",
                params={"session_id": session_id},
            )
            == "['P1','P2']\t['P1','P2']\t['default','P3','P4','P5','P1','P2']\n"
        )
E       assert "['P1','P2']\t['default','P3','P5','P1','P2']\t['default','P3','P4','P5','P1','P2']\n" == "['P1','P2']\t['P1','P2']\t['default','P3','P4','P5','P1','P2']\n"
E         - ['P1','P2']	['P1','P2']	['default','P3','P4','P5','P1','P2']
E         + ['P1','P2']	['default','P3','P5','P1','P2']	['default','P3','P4','P5','P1','P2']
E         ?            	 ++++++++++++++++++++
```

**Comment by Enmk** [comment]
# Errors analysis for [2684fc5b](https://github.com/ClickHouse/ClickHouse/commit/2684fc5b1b31a00d28c31d6182fbb3597ce45747/status-details)
### [ClickHouse Integration Tests (asan) [1/6]](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__asan__[1/6].html)
 - [test_keeper_zookeeper_converter/test.py::test_acls[False]](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__asan__[1/6]/integration_run_parallel1_0.log) - timed out ZK connection
```python
/usr/local/lib/python3.8/dist-packages/kazoo/client.py:635: KazooTimeoutError
```
 - [`test_keeper_map/test.py::test_create_drop_keeper_map_concurrent`](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__asan__[1/6]/integration_run_parallel1_0.log) - Timeout
```
E               Failed: Timeout >900.0s
```

### [ClickHouse Integration Tests (asan) [5/6]](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__asan__[5/6].html)

- [`test_settings_profile/test.py::test_function_current_profiles`](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__asan__[5/6]/integration_run_parallel2_0.log) `currentProfiles()` now reports implicitly applied profiles too (e.g. `P5` via `ROLE`), which I believe is more correct behavior. Test needs to be updated though:
```
>       assert (
            instance.http_query(
                "SELECT defaultProfiles(), currentProfiles(), enabledProfiles()",
                user="robin",
                params={"session_id": session_id},
            )
            == "['P1','P2']\t['P1','P2']\t['default','P3','P4','P5','P1','P2']\n"
        )
E       assert "['P1','P2']\t['default','P3','P5','P1','P2']\t['default','P3','P4','P5','P1','P2']\n" == "['P1','P2']\t['P1','P2']\t['default','P3','P4','P5','P1','P2']\n"
E         - ['P1','P2']	['P1','P2']	['default','P3','P4','P5','P1','P2']
E         + ['P1','P2']	['default','P3','P5','P1','P2']	['default','P3','P4','P5','P1','P2']
E         ?            	 ++++++++++++++++++++
```

### [ClickHouse Integration Tests (release) [3/4]](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__release__[3/4].html)
- [`test_settings_profile/test.py::test_function_current_profiles`](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__release__[3/4]/integration_run_parallel3_0.log) - same as above

### [ClickHouse Integration Tests (tsan)](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__tsan__[1/6].html)
- [`test_hedged_requests/test.py::test_receive_timeout2`](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__tsan__[1/6]/integration_run_parallel1_0.log)
```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 209, stderr: Received exception from server (version 22.12.1):
E           Code: 209. DB::Exception: Received from 172.16.8.2:9000. DB::NetException. DB::NetException: Timeout exceeded while reading from socket (node_1:9000, receive timeout 3000 ms): While executing Remote. Stack trace:
...
```

### [ClickHouse Integration Tests (tsan) [5/6]](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__tsan__[5/6].html)
- [`test_settings_profile/test.py::test_function_current_profiles`](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/integration_tests__tsan__[5/6]/integration_run_parallel2_0.log) - same as other `test_settings_profile/test.py::test_function_current_profiles` cases

### [ClickHouse Stress Test (asan)](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/stress_test__asan_.html)
- Seems unrelated: https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/stress_test__asan_/fatal_messages.txt
```<Fatal> : Logical error: 'Part all_4_4_0 doesn't exist'```


### [ClickHouse Stress Test (debug)](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/stress_test__debug_.html)
Looks like [unrelated OOM ](https://s3.amazonaws.com/clickhouse-test-reports/42641/2684fc5b1b31a00d28c31d6182fbb3597ce45747/stress_test__debug_/dmesg.log)


# Conclusion
Need to modify [`tests/integration/test_settings_profile/test.py::test_function_current_profiles`](https://github.com/ClickHouse/ClickHouse/blob/127c980aff4cdcc9c5959cce79fec6704aa2665f/tests/integration/test_settings_profile/test.py#L536) to reflect fixed behavior of `currentProfiles()`

File: src/Access/SettingsProfilesInfo.cpp [hunk 2]
```diff
@@ -55,4 +62,26 @@ SettingsProfilesInfo::getConstraintsAndProfileIDs(const std::shared_ptr<const Se
...
+            result.push_back(p->second);
+        else
+        {
+            if (const auto name = access_control.tryReadName(profile_id))
+                // We could've updated cache here, but it is a very rare case, so don't bother.
+                result.push_back(*name);
+            else
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Unable to get profile name for {}", toString(profile_id));
+        }
+    }
```
Comment (tavplubix) [inline src/Access/SettingsProfilesInfo.cpp:82]: @Enmk, what does it fix and how? Looks like it simply replaces `std::out_of_range` with `LOGICAL_ERROR` without fixing the root casue

Comment (rschu1ze) [reply src/Access/SettingsProfilesInfo.cpp:82]: #54656

Comment (vitlibar) [reply src/Access/SettingsProfilesInfo.cpp:82]: @Enmk `SettingsProfilesInfo` must be internally consistent, so I think if we fix that consistency in https://github.com/ClickHouse/ClickHouse/pull/57263 then we only need here a logical error:

```
    for (const auto & profile_id : profiles)
    {
        const auto p = names_of_profiles.find(profile_id);
        if (p != names_of_profiles.end())
            result.push_back(p->second);
        else
            throw Exception(ErrorCodes::LOGICAL_ERROR, "Unable to get profile name for {}", toString(profile_id));
    }
```

**Pull request merged**